### PR TITLE
feat(user): run GDPR deletion as a durable admin queue

### PR DESCRIPTION
**Pull request by eshurakov**
## Summary

Replace one-shot Cloud GDPR deletion with a durable admin deletion queue.

- Disable matching Cloud accounts immediately on enqueue.
- Persist a deletion-task catalog in PostgreSQL and drive teardown through cron retries.
- Anonymize only after required tasks succeed or an admin marks a stuck task done.
- Treat deletion-in-progress as gone for owner matching, and keep HMAC/encrypted audit subjects from re-enqueuing already-covered resources.
- Scrub usage prompt prefixes as a paginated queue step so large usage histories cannot timeout deletion.
- Add `/admin/deletion-queue` for intake, retry, and mark-done.
- Signup/account-link still block on a real in-flight deletion, but unexpected fence lookup errors fail open and are reported to Sentry.

PostHog, Substack, and Pylon remain CSA-handled for this first cut. Add-requests is hidden in the queue UI; start deletion from the user profile or CSA API.

## Test plan

- [x] Enqueue a known Cloud user from the user admin GDPR form and confirm the request appears on `/admin/deletion-queue`.
- [x] Confirm the request persists with the v1 catalog and usage-prompt cleanup is a teardown step.
- [x] Run `process-user-deletions` until teardown succeeds and anonymize runs only after required tasks complete.
- [x] Retry a stuck provider task from the admin UI; mark-done still requires evidence.
- [x] Cancel a pending request; the email is redacted and the summary count updates after invalidate.
- [x] Soft-delete still pages usage prompt prefixes before anonymizing.

Local E2E on `http://localhost:6200/` (offset 3200):

- Fake-admin login can open `/admin/deletion-queue`. Start deletion from user Admin Tools (handbook checkbox → confirm → queued).
- Queue list, email filter, detail sheet, dedicated request page, cancel, and re-queue after cancel all work.
- GDPR confirm no longer hydrates a `<div>` inside `<p>`. Cancel/retry now invalidate the queue summary.
- Existing-user re-login is not fenced; the fence only blocks new signup / account link. Unexpected fence errors fail open.
- Chunked usage: 2500 prompt-prefix rows, batch 1000 → continue, continue, succeeded, then anonymize.
- Chunked CLI sessions: 25 `cli_sessions_v2` rows, batch 10. Invalid IDs (`ses_chunk_001`) returned `http_400` because ingest requires `ses_` + 26 chars. Valid IDs + admin retry → continue, continue, succeeded.
- Customer.io succeeded locally. KiloClaw and CLI v1 were not applicable on these users.

Not covered locally: PostHog/Substack/Pylon (CSA).

### CODE REVIEW

File: apps/web/src/lib/user/deletion-queue/deletion-enqueue.ts [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+        targetEmailHmac: writeHmac.hmac,
+        targetEmailHmacKeyVersion: writeHmac.version,
+        subjectKey: 'request',
+        details: { catalog_version: USER_DELETION_CATALOG_VERSION },
+      });
+
+      return { status: 'enqueued' as const, requestId: request.id };
+    });
+  } catch (error) {
+    if (isUniqueActiveTargetError(error)) {
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-enqueue.ts:204]: Fixed. Unique-violation recovery now falls back to `findActiveRequestByUserId` after the email HMAC lookup.

File: apps/web/src/lib/user/deletion-queue/deletion-outcomes.ts [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+          .where(eq(user_deletion_requests.id, params.requestId));
+        return true;
+      }
+      return false;
+    }
+    if (step.status !== UserDeletionStepStatus.Running) {
+      return false;
+    }
+
+    const [request] = await tx
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-outcomes.ts]: Fixed. `persistHandlerOutcomeTx` now locks the request row before the step row.

File: apps/web/src/lib/user/deletion-queue/deletion-worker.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  while (wave.length < USER_DELETION_MAX_CONCURRENT_TASKS) {
+    const remainingMs = deadlineAt - Date.now();
+    if (remainingMs < USER_DELETION_STOP_STARTING_RESERVE_MS) break;
+
+    const request = await selectEligibleDeletionRequest({
+      excludeRequestIds: usedRequestIds,
+    });
+    if (!request) {
+      if (wave.length === 1) {
+        const same = await selectTaskItem(usedRequestIds[0], remainingMs, usedStepKeys, usedScopes);
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-worker.ts]: Fixed. Same-request fill now uses `wave[0].requestId`.

File: apps/web/src/lib/user/deletion-queue/deletion-worker.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+      }
+      const requestIds = [...new Set(wave.map(item => item.requestId))];
+      for (const requestId of requestIds) {
+        await advanceDeletionGates(requestId);
+      }
+    }
+    await sweepUnclaimableDeletionGates(deadlineAt);
+    await finishWorkerLease(ownerToken, UserDeletionWorkerOutcome.Success, region);
+    return { outcome: UserDeletionWorkerOutcome.Success, processed };
+  } catch {
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-worker.ts]: Fixed. Worker failures are captured with `captureException` before finishing the lease.

File: apps/web/src/lib/user/deletion-queue/handlers/kiloclaw.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+  for (const instance of batch) {
+    const reserve = continueIfLowTime(context, progress);
+    if (reserve) return reserve;
+
+    let marked;
+    try {
+      marked = await markActiveInstanceBatchDestroyedForGdpr(userId, [instance.id]);
+    } catch {
+      return {
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/handlers/kiloclaw.ts]: Fixed. `classifyKiloclawMarkError` maps ownership/empty-batch mismatches to `needs_attention` and other errors to `retry`.

File: apps/web/src/lib/user/deletion-queue/deletion-request-selector.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    .from(user_deletion_steps)
+    .innerJoin(
+      user_deletion_requests,
+      eq(user_deletion_steps.request_id, user_deletion_requests.id)
+    )
+    .where(
+      and(
+        isNull(user_deletion_requests.paused_at),
+        inArray(user_deletion_requests.status, [
+          UserDeletionRequestStatus.Pending,
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-request-selector.ts]: Deleted unused `hasDueDeletionWork`. Worker eligibility already skips attention-blocked Pending via `selectEligibleDeletionRequest`.

File: apps/web/src/lib/user/deletion-queue/handlers/posthog.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        return { kind: 'needs_attention', errorCode: 'posthog_ambiguous' };
+      }
+
+      personIds = [...new Set(persons.map(person => person.id))];
+      const distinctIds = [...new Set(persons.flatMap(person => person.distinctIds))];
+
+      const lostBeforeReserve = await assertCurrentClaim(context);
+      if (lostBeforeReserve) return lostBeforeReserve;
+
+      if (!effect) {
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/handlers/posthog.ts]: Leaving as-is. The writeback gap is real if the person set changes between reserve and a later poll-only run, but PostHog is not in the live catalog (CSA still owns it). Same-run poll uses the recomputed IDs. Will persist `encrypted_resource_ids` on submit if the step is re-enabled.

File: apps/web/src/app/admin/api/users/gdpr-removal/route.ts [hunk 8]
```diff
@@ -1,16 +1,55 @@
...
 
-export async function POST(request: NextRequest): Promise<NextResponse<GdprRemovalResponse>> {
-  const { user: adminUser, authFailedResponse } = await getUserFromAuth({ adminOnly: true });
+export async function GET(request: NextRequest): Promise<NextResponse<GdprRemovalGetResponse>> {
+  const { authFailedResponse } = await getUserFromAuth({ adminOnly: true });
+  if (authFailedResponse) return authFailedResponse;
+
+  const requestId = request.nextUrl.searchParams.get('requestId');
+  if (requestId) {
+    const deletion = await getUserDeletionRequestById(requestId);
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/app/admin/api/users/gdpr-removal/route.ts:27]: Leaving as-is. Admin-only GET; `UserAdminGdprRemoval` only polls with a DB UUID from enqueue. A curl typo is a 500, not a user-facing path.

File: apps/web/src/app/api/internal/support/users/gdpr-removal/route.ts [hunk 9]
```diff
@@ -85,30 +101,121 @@ export async function POST(request: NextRequest): Promise<NextResponse> {
...
+  }
+
+  const csaRequestIdParsed = RequestIdSchema.safeParse(request.headers.get('x-request-id'));
+  if (!csaRequestIdParsed.success) {
+    return NextResponse.json({ error: 'Invalid requestId header' }, { status: 400 });
+  }
+  const csaRequestId = csaRequestIdParsed.data;
+
+  const requestId = request.nextUrl.searchParams.get('requestId');
+  if (!requestId) {
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/app/api/internal/support/users/gdpr-removal/route.ts]: Fixed. GET now validates `requestId` with `z.string().uuid()` and returns 400 without hitting the DB.

File: apps/web/src/app/admin/api/users/deletion-steps/retry/route.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+function parseRetryBody(
+  body: unknown
+):
+  | { kind: 'preflight'; requestId: string; reason: string }
+  | { kind: 'step'; requestId: string; stepKey: UserDeletionStepKey; reason: string }
+  | null {
+  if (typeof body !== 'object' || body === null) return null;
+  if (!('requestId' in body) || !('reason' in body)) return null;
+  if (typeof body.requestId !== 'string' || body.requestId.length === 0) return null;
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/app/admin/api/users/deletion-steps/retry/route.ts]: Deleted the unused REST route. Live retry is tRPC `retryTask` / `retryPreflight`, already UUID-validated.

File: apps/web/src/routers/admin/user-deletion-queue-router.ts [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+}
+
+function decodeCursor(cursor: string): { createdAt: string; id: string } | null {
+  try {
+    const decoded = Buffer.from(cursor, 'base64url').toString('utf8');
+    const split = decoded.indexOf('\t');
+    if (split <= 0) return null;
+    const createdAt = decoded.slice(0, split);
+    const id = decoded.slice(split + 1);
+    if (!createdAt || !z.string().uuid().safeParse(id).success) return null;
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/routers/admin/user-deletion-queue-router.ts:136]: Leaving as-is. The value is parameterized, not injectable. Encoded `createdAt` is raw Postgres timestamptz text, so `z.string().datetime()` would reject real cursors. Failed decode already drops the cursor and returns page 1.

File: ENVIRONMENT.md [hunk 12]
```diff
@@ -36,6 +36,12 @@ Manage shared web env var additions and rotations with `pnpm web:env set <VARIAB
 
 - `NEXT_PUBLIC_POSTHOG_KEY` - PostHog public API key for client-side analytics; used in `apps/web/src/components/PostHogProvider.tsx`. [PUBLIC]
 - `NEXT_PUBLIC_POSTHOG_DEBUG` - Enables PostHog debug logging; checked in `apps/web/src/components/PostHogProvider.tsx` and `apps/web/src/lib/stytch.ts`. [PUBLIC]
+- `POSTHOG_PERSONAL_API_KEY` - PostHog personal API key used by user-deletion PostHog cleanup. `[SECRET]`
+- `POSTHOG_ENVIRONMENT_ID` - PostHog project/environment ID used by user-deletion PostHog cleanup. [SERVER]
+- `POSTHOG_HOST` - Optional PostHog API host override for user-deletion cleanup; defaults to `https://us.posthog.com`. [SERVER]
+- `PYLON_API_KEY` - Pylon REST API key used by user-deletion contact and reply cleanup. `[SECRET]`
+- `PYLON_HOST` - Optional Pylon API host override for local user-deletion cleanup; defaults to `https://api.usepylon.com`. [SERVER]
+- `CUSTOMERIO_TRACK_BASE` - Optional Customer.io Track API base override for local user-deletion cleanup; defaults to `https://track.customer.io`. [SERVER]
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply ENVIRONMENT.md:45]: Added `SUBSTACK_PUBLICATION_URL` to `ENVIRONMENT.md`. Not adding it to `.env.local.example` — sibling mock-injected hosts (`POSTHOG_HOST`, `PYLON_HOST`, `CUSTOMERIO_TRACK_BASE`) are not there either.

File: .env.local.example [hunk 13]
```diff
@@ -103,9 +103,7 @@ STRIPE_KILOCLAW_2026_05_10_COMMIT_PRICE_ID=price_test_kiloclaw_2026_05_10_commit
 # ============================================================================
 # Internal API secret (generate: openssl rand -base64 32)
 INTERNAL_API_SECRET=changeme
-# Support service API (CSA GDPR web only). Distinct from INTERNAL_API_SECRET.
-# Leak = enumerate any email + wipe any non-staff, non-live-sub customer.
-# Generate: openssl rand -base64 32
+# Customer Support Automation internal API bearer token (generate: openssl rand -base64 32)
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply .env.local.example:111]: Updated. Leak can still look up emails and enqueue deletion; access disable is deferred to worker preflight and pending requests can be cancelled. Restored that wording in `ENVIRONMENT.md` and `.env.local.example` instead of the old wipe-immediately warning.

File: apps/web/src/lib/user/deletion-queue/deletion-enqueue.test.ts [hunk 14]
```diff
@@ -0,0 +1,468 @@
...
+      targets: [{ email: 'DUP@example.com' }],
+    });
+    expect(first[0]?.status).toBe('enqueued');
+    expect(second[0]).toMatchObject({
+      status: 'already_active',
+      requestId: first[0] && first[0].status === 'enqueued' ? first[0].requestId : undefined,
+    });
+  });
+
+  it('returns already_active when unique-violation recovery finds the request by user_id', async () => {
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-enqueue.test.ts]: Renamed the test to match the in-tx user_id path it actually covers. Catch-path comment notes the remaining race: user_id unique index across two different email advisory locks. Not deleting the catch; a sequential test cannot reach it without hiding the user_id row.

File: apps/web/src/lib/user/deletion-queue/deletion-catalog.ts [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+  //   dependsOn: [],
+  //   allowsManualVerification: true,
+  // },
+  {
+    stepKey: UserDeletionStepKey.Anonymize,
+    phase: UserDeletionPhase.Anonymize,
+    cooldownScope: null,
+    maxOrdinaryAttempts: USER_DELETION_MAX_ORDINARY_ATTEMPTS,
+    dependsOn: TEARDOWN_KEYS,
+    allowsManualVerification: true,
```
Comment (pandemicsyn) [inline apps/web/src/lib/user/deletion-queue/deletion-catalog.ts]: bot: **Blocking: manual verification can complete deletion without anonymizing the account**

`markTaskManuallyVerified` treats this step as successful and sets `anonymized_at`, but it does not invoke `anonymizeCloudUserData`. Because the completion gate accepts manually verified catalog steps, an administrator can clear an attention-blocked anonymization task and complete the request while the Cloud user row and its PII have not been scrubbed.

Do not permit manual verification for `UserDeletionStepKey.Anonymize` (`allowsManualVerification: false`), or make the manual action execute and verify `anonymizeCloudUserData` atomically. Please add a regression test showing that manual verification is rejected and cannot advance the request.

File: apps/web/src/lib/user/deletion-queue/deletion-preflight.ts [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+export async function runDeletionPreflight(requestId: string): Promise<PreflightResult> {
+  return db.transaction(async tx => {
+    const [request] = await tx
+      .select()
+      .from(user_deletion_requests)
+      .where(eq(user_deletion_requests.id, requestId))
+      .for('update');
+
+    if (!request) return { kind: 'skipped', reason: 'not_found' };
+    if (request.status !== UserDeletionRequestStatus.Pending) {
```
Comment (pandemicsyn) [inline apps/web/src/lib/user/deletion-queue/deletion-preflight.ts]: bot: **Blocking: a pause can be bypassed by an already-selected worker wave**

The worker selects a request into a wave in one transaction, then runs preflight later. If an administrator pauses the request in that interval, this guard still sees a `Pending` request because it does not re-check `paused_at`. Preflight can therefore disable the user's access and promote a paused request to `InProgress`, despite the admin control intended to stop it.

After locking the request, return a skipped result when `request.paused_at !== null` (ideally with a distinct `paused` reason). Please add a concurrency regression test that pauses a request after wave selection but before `runDeletionPreflight`, and asserts that neither access nor request status changes.

**Review by pandemicsyn — APPROVED** [review]
two findings from my local robot inline. Both sound very edge cases (especially the pause race) , so feel free to ignore.

File: apps/web/src/lib/user/deletion-queue/deletion-catalog.ts (hunk 15 above)
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-catalog.ts]: Fixed. Anonymize is no longer manually verifiable (`allowsManualVerification: false`). `markTaskManuallyVerified` rejects it and no longer stamps `anonymized_at`; the request stays pending.

File: apps/web/src/lib/user/deletion-queue/deletion-preflight.ts (hunk 16 above)
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/deletion-preflight.ts]: No longer applicable. Pause was removed end-to-end; the queue only has active and terminal request states, so there is no `paused_at` to re-check after wave selection.

File: packages/db/src/migrations/0217_tidy_vulcan.sql [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+CREATE INDEX "IDX_user_deletion_requests_fairness" ON "user_deletion_requests" USING btree ("last_progress_at","created_at","id") WHERE "user_deletion_requests"."status" IN ('pending', 'in_progress', 'finalizing');--> statement-breakpoint
+CREATE INDEX "IDX_user_deletion_requests_email_hmac" ON "user_deletion_requests" USING btree ("target_email_hmac","target_email_hmac_key_version");--> statement-breakpoint
+CREATE INDEX "IDX_user_deletion_requests_user_id" ON "user_deletion_requests" USING btree ("user_id");--> statement-breakpoint
+CREATE INDEX "IDX_user_deletion_steps_due" ON "user_deletion_steps" USING btree ("status","available_at","id") WHERE "user_deletion_steps"."status" IN ('pending', 'retry_wait', 'running');--> statement-breakpoint
+CREATE INDEX "IDX_user_deletion_steps_request_id" ON "user_deletion_steps" USING btree ("request_id");--> statement-breakpoint
+CREATE INDEX "IDX_kilocode_users_lower_google_user_email" ON "kilocode_users" USING btree (lower("google_user_email"));
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply packages/db/src/migrations/0217_tidy_vulcan.sql]: No longer applicable. The unused `lower(google_user_email)` index was dropped; this migration file is gone.

File: apps/web/src/lib/user/deletion-queue/handlers/kiloclaw.ts [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+    if (reserve) return reserve;
+
+    let marked;
+    try {
+      marked = await markActiveInstanceBatchDestroyedForGdpr(userId, [instance.id]);
+    } catch (error) {
+      return classifyKiloclawMarkError(error, resourceHmac(instance.id));
+    }
+
+    const afterMark = continueIfLowTime(context, progress);
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/deletion-queue/handlers/kiloclaw.ts]: Fixed. Removed the post-mark `continueIfLowTime` after `markActiveInstanceBatchDestroyedForGdpr`. Destroy still has its own abort; the instance is no longer hidden from later runs before teardown.

File: apps/web/src/lib/user/gdpr-removal.ts [hunk 19]
```diff
@@ -23,65 +14,28 @@ export async function performGdprRemoval(
...
-        extra: { userId, instanceIds: group.instanceIds },
-      });
-    }
+  const [result] = await enqueueUserDeletionTargets({
+    actor: { kiloUserId: options.actor.id, email: options.actor.email },
+    targets: [
+      {
+        email: user.google_user_email,
+        trustedUserId: userId,
+        allowSelf: true,
```
[the remark this answers is not in the document]
Comment (eshurakov) [reply apps/web/src/lib/user/gdpr-removal.ts:28]: Fixed in c1ec9dfb0.

`protected_self` is an intake rule. Preflight no longer re-runs it against the requester, so an email-confirmed self-service request can promote and disable access. Admin self-delete is still refused at enqueue. Bot and subscription checks are unchanged.

Left `{ status: 'deleted' }` on enqueue for the shipped mobile client.

File: apps/web/src/lib/user/index.ts [hunk 20]
```diff
@@ -821,6 +835,15 @@ export async function linkAccountToExistingUser(
   const existingUser = await findUserById(existingKiloUserId);
   if (!existingUser) return failureResult('USER-NOT-FOUND');
 
+  if (
```
Comment (pandemicsyn) [inline apps/web/src/lib/user/index.ts:838]: bot: This fence protects the normal linking path, but `createOrUpdateUser` has a separate same-email WorkOS branch that deletes/reinserts `user_auth_provider` and rotates `web_session_pepper` without calling `linkAccountToExistingUser` or `authPassesDeletionFence`. An active deletion request can therefore still have its auth identity mutated through a WorkOS login. Please apply the same fence in that transaction (or route WorkOS through this helper) and add an active-deletion + WorkOS auto-link regression test.

**Review by pandemicsyn — APPROVED** [review]
one inline, seems very edge case though. If a user logs in via work os after having requested deletion at the exact time the deletion happens ?  I'm not sure its worth the effort.
